### PR TITLE
Improve Binary Search

### PR DESCRIPTION
**Pull request by glebkema**
I didn't сombine the commits because I used them to group and explain my changes. If necessary, I will redo the PR.

1. toughen up tests
    - check more missing and present values
    - check triple value search (and let the second group of tests not repeat the first one)
    - check search for null in empty array (to make sure the algorithm distinguishes zero from emptiness)
2. optimize the algorithm
    - use `return` to stop the loop and simplify the `if then` structure
    - sort the array on initialization, not before each search

- [x] tests passed successfully
- [x] maven `compile` executed

### CODE REVIEW

File: src/main/java/com/pd/algorithms/BinarySearch.java [hunk 1]
```diff
@@ -8,28 +8,28 @@ public class BinarySearch {
...
     this.arr = arr;
     this.size = arr.length;
   }
 
   public boolean search(int num) {
-    Arrays.sort(arr);
     return binarySearch(arr, 0, size - 1, num);
   }
 
   public boolean binarySearch(int[] array, int low, int high, int value) {
```
Comment (Pradyuman7) [inline src/main/java/com/pd/algorithms/BinarySearch.java:20]: I guess we can make this method `private` since this is mainly to be called from the `search` public method. Could you please update this and also update the docs, if needed?

**Comment by glebkema** [comment]
Ok, I will do it in about 12 hours

Pradyuman Dixit ***@***.***> schrieb am Di., 11. Okt. 2022,
09:20:

**Comment by glebkema** [comment]
I realized that our recursion does not require a `while` loop. So I replaced it with an `if` statement and reordered all the `if` statements to make the code clearer. 

Also I added some more tests for negative values and zero in the array.

**Review by Pradyuman7 — APPROVED** [review]
LGTM 💯
